core/schema/module.go

```diff
@@ -316,176 +333,841 @@ func (s *moduleSchema) functionWithArg(ctx context.Context, fn *core.Function, a
+	newDepSrc.Self = newDepSrc.Self.Clone()
+
+	depSourceSubpath, err := depSrc.Self.Subpath()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get module source path: %w", err)
+	}
+	if escapes, err := escapesParentDir(rootPath, depSourceSubpath); err != nil {
+		return nil, fmt.Errorf("failed to check if module source path escapes root: %w", err)
+	} else if escapes {
+		return nil, fmt.Errorf("module dep source path escapes root: %q", depSourceSubpath)
+	}
+	fullDepSourcePath := filepath.Join(rootPath, depSourceSubpath)
+
+	switch src.Self.Kind {
+	case core.ModuleSourceKindGit:
+		newDepSrc.Self.AsGitSource.Value.Subpath = fullDepSourcePath
+
+		// preserve the git metadata by just constructing a modified git source ref string
+		// and using that to load the dep
+		newDepRefStr, err := newDepSrc.Self.RefString()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get module source ref string: %w", err)
+		}
+
+		err = s.dag.Select(ctx, s.dag.Root(), &newDepSrc,
+			dagql.Selector{
+				Field: "moduleSource",
+				Args: []dagql.NamedInput{
+					{Name: "refString", Value: dagql.String(newDepRefStr)},
+					{Name: "stable", Value: dagql.Boolean(true)},
+				},
+			},
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load git dep: %w", err)
+		}
+		return newDepSrc.Self, nil
+
+	case core.ModuleSourceKindLocal:
+		newDepSrc.Self.AsLocalSource.Value.Subpath = fullDepSourcePath
+
+		err = s.dag.Select(ctx, s.dag.Root(), &newDepSrc,
+			dagql.Selector{
+				Field: "moduleSource",
+				Args: []dagql.NamedInput{
+					{Name: "refString", Value: dagql.String(fullDepSourcePath)},
+					{Name: "rootDirectory", Value: dagql.Opt(dagql.NewID[*core.Directory](src.Self.RootDirectory.ID()))},
+				},
+			},
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load git dep: %w", err)
+		}
+		return newDepSrc.Self, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported module source kind: %q", src.Self.Kind)
 	}
-	return deps.TypeDefs(ctx)
 }
 
-func (s *moduleSchema) functionCallReturnValue(ctx context.Context, fnCall *core.FunctionCall, args struct {
-	Value core.JSON
-}) (dagql.Nullable[core.Void], error) {
-	// TODO: error out if caller is not coming from a module
-	return dagql.Null[core.Void](), fnCall.ReturnValue(ctx, args.Value)
+func (s *moduleSchema) moduleSourceAsModule(
+	ctx context.Context,
+	src dagql.Instance[*core.ModuleSource],
+	args struct{},
+) (*core.Module, error) {
+	if src.Self.RootDirectory.Self == nil {
+		return nil, fmt.Errorf("cannot load module from module source with no root directory")
+	}
+
+	var mod dagql.Instance[*core.Module]
+	err := s.dag.Select(ctx, s.dag.Root(), &mod,
+		dagql.Selector{
+			Field: "module",
+		},
+		dagql.Selector{
+			Field: "withSource",
+			Args: []dagql.NamedInput{
+				{Name: "source", Value: dagql.NewID[*core.ModuleSource](src.ID())},
+			},
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load local src module: %w", err)
+	}
+	return mod.Self, nil
```

Comment:
Does this mean `ModuleSource.asModule` returns an uninitialized module, while `Directory.asModule` returns an initialized one?

Not a dealbreaker, just looking for any inconsistency in our vocabulary, if there's a better wording in either case it seems worth doing, but no big deal

Reply:
Yeah good point, updated `Directory.asModule` to also not initialize for consistencies sake.

Reply:
In the current API, that would mean `Module.withSource` needs to return an initialized module, which doesn't work because we explicitly want it to be a possibly completely uninitialized module (as in, it may have not even been configured w/ a name/sdk/etc. yet). That also ties into the fact that we want to be able to get the generated source for the Module even if it doesn't yet compile itself.

I could see this being doable if `ModuleSource` had both an `asInitializedModule` and an `asUninitializedModule` field. But at that point it seems extraneous since callers can choose whether to chain `initialize` on the `Module`.

Basically, I don't really mind `initialize` being something explicit since ultimately it can't actually be hidden if it's in the API and is useful in that it lets you use a `Module` without it actually being fully configured and/or fully compiling yet. I'm also more okay with breaking API changes here than elsewhere here because this whole API is still part of our "experimental/preview" module support and afaik the consumers are the CLI+daggerverse.

---

That being said, one thing to consider is to maybe just remove `Directory.asModule`. It's technically fairly redundant since it internally just provides the `Directory` and the `subpath` arg to `ModuleSource` and then goes through those paths, so I'm not sure whatever convenience it offers is actually significant to justify its existence anymore. I'll leave it as is until I hear back, but let me know what you think.

Reply:
That all makes sense to me, this just felt like a super subtle shift, so wanted to be super explicit about it being a conscious choice. :+1: 

I slightly prefer leaving `Directory.asModule` around as a convenience, since it's more ergonomic when you're playing around in the GraphQL playground, since you can just chain instead of having to pass in a `DirectoryID`